### ISSUE TITLE
Differentiability of proxy T_60 error

### ISSUE THREAD

**Issue by SAGNIKMJR** [nikhilsinghmus/room2reverb#29]
Hi Nikhil,

Thanks for open-sourcing the code!

I have a question regarding the differentiability of the [proxy T_60 error](https://github.com/nikhilsinghmus/image2reverb/blob/f855f49d50e7e0d7830e114cc46d54a711e4d5f2/image2reverb/model.py#L71). Firstly, in the in `compare_t60` in [image2reverb.util](https://github.com/nikhilsinghmus/image2reverb/blob/main/image2reverb/util.py), `detach` is being called on both predictions and ground truths, which would make any further computation non-differentiable. Secondly, inside `estimate_t60`, which is being called from `compare_t60`, there exists a bunch of `torch.where` operations, which are again non-differentiable. So it looks like the proxy T_60 error won't be differentiable and upon adding it to the other losses, shouldn't really change the model training as the model won't have any control over it.

However, in Sec. 4.2 (Ablation Study) of the paper, it looks like not having the proxy T_60 error term in the objective is creating a difference. Am I missing something here?

Best,
Sagnik

**Comment by nikhilsinghmus** [comment]
Hi Sagnik,

Thanks for the question! This is indeed puzzling, looking at it now. It might take me a little while to reproduce, compare codebases, and investigate. The ablation we ran definitely showed a difference (more than once), so hopefully this isn't an error, but will work on tracing it.

**Comment by nikhilsinghmus** [comment]
Hi @SAGNIKMJR, thank you again for pointing this out, and for your patience! A quick update from my side: I think the calls to .detach() aren't supposed to be there, but of course this still leaves the problem of the index tensors.

My sincerest apologies for this error! I think, in retrospect, a more rigorous experimental design would have caught this sooner, and brought our attention to the differentiability issue. I appreciate your diligence, once again.

I managed to track down versions of our ablation models (same conditions as the reported ablations, if not the identical checkpoints), and also run another quick round of training, and you're right! The difference between them is quite small: tested across 5 runs with matched random seeds. I think, after looking into it, we may have had the wrong data for the T60p ablation when generating that plot (e.g. another no-depth checkpoint, since they are close), or the difference was due to another source of variation. Unfortunately, I can't perfectly reproduce the experiments to confirm, but I think this is most likely.

I'm working on a "softer" differentiable proxy T60 function to see if we can get a further performance boost. So far, I'm not seeing a consistent improvement, but will update once I've had a chance to work on this more. Again, thank you for your insight.

**Comment by SAGNIKMJR** [comment]
No problem. Thanks for taking out the time to look into it!

**Comment by bjut-chunxiwang** [comment]
Have there been any recent updates or improvements to the T60 component of the loss function?

**Comment by nikhilsinghmus** [comment]
Hi @bjut-chunxiwang, thanks for the message. I prototyped some updates back then, but ultimately I think the best resource is @SAGNIKMJR's [NeurIPS 2022](https://arxiv.org/abs/2206.04006) paper.

**Comment by bjut-chunxiwang** [comment]
Thank you very much for your prompt reply. By the way, have you ever tried computing the DRR error? On my side, I obtained DRR mean (%): -212.39 | std (%): 331.89. I wonder if your results are similar. It seems that estimating DRR purely from images is not theoretically sound. I’d like to ask whether incorporating speech or other audio information would make these evaluation metrics more interpretable.

**Comment by nikhilsinghmus** [comment]
@bjut-chunxiwang yes you're right, in the single-image setup, we focus on late reverberation (as discussed in the paper). Since DRR depends on the source and listener positions, it isn't identified in this kind of data. One strategy could be to use synthetic data: simulate different DRRs within the same scene. With that, you could either model the RIR conditional on the DRR, or try to get images which show the source and listener positions so they already contain DRR-relevant information.
